Add repo setup

ElePT opened the pull request:
Add setup for automatic checks (lint, black, spelling, copyright, unit tests) using github actions/workflows.

Reference: https://github.com/qiskit-community/qiskit-machine-learning.

**Note**:

- For simplicity, as we don't have docs yet, all of the docs-related actions have been left out of this PR, and saved for a follow-up
- Items to be discussed (but not included in this PR):
   - mypy
   - Neko integration tests 
   - tox vs. make
   - copyright and spelling
   - docs + docs infrastructure
   
**Missing**:

- [x] secrets.encrypted_rclone_key
- [x] secrets.encrypted_rclone_iv
- [x] secrets.TWINE_PASSWORD
- [x] secrets.GITHUB_TOKEN

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/qiskit-community/qiskit-algorithms?pullRequest=1) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/qiskit-community/qiskit-algorithms?pullRequest=1) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/qiskit-community/qiskit-algorithms?pullRequest=1) it.</sub>

ElePT commented on .editorconfig:
```diff
@@ -0,0 +1,51 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2017, 2022.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
```
I am not sure what to do with the license. Pending discussion I guess.

woodsp-ibm commented on .github/ISSUE_TEMPLATE/BUG_REPORT.yaml:
```diff
@@ -0,0 +1,50 @@
+      # in easier.  The type is 'textarea' rather than three separate 'input's
+      # to make the resulting issue body less noisy with headings.
+      value: |
+        - **Qiskit Machine Learning version**:
```
```suggestion
        - **Qiskit Algorithms version**:
```

woodsp-ibm commented on .github/workflows/deploy-code.yml:
```diff
@@ -0,0 +1,39 @@
+          python-version: ${{ matrix.python-version }}
+      - name: Deploy to Pypi
+        env:
+          TWINE_PASSWORD: ${{ secrets.TWINE_PASSWORD }}
```
We will need secrets setup to work evidently.

woodsp-ibm commented on .github/workflows/deploy-docs.yml:
```diff
@@ -0,0 +1,83 @@
+          make html
+          tools/deploy_documentation.sh
+        shell: bash
+  deploy-translatable-strings:
```
Translatable strings needs further setup if we are going to have this - it was done for the apps of course - not sure if its going to persist there or not

woodsp-ibm commented on .github/workflows/main.yml:
```diff
@@ -0,0 +1,313 @@
+        shell: bash
+      - name: Combined Deprecation Messages
+        run: |
+          sort -f -u /tmp/u38/ml.dep /tmp/u39/ml.dep /tmp/u310/ml.dep /tmp/u311/ml.dep /tmp/m38/ml.dep /tmp/m311/ml.dep /tmp/w38/ml.dep /tmp/w311/ml.dep || true
```
This is more just a comment. I do not know why the names of these files were not a little more generic. The `ml` aspect is because its machine learning (if you look at the same file in the other app repos the name reflects the app - maybe its in case it was all run in one place the names would not collide). I guess it could be changed to `alg.dep` and `alg.dat` below and elsewhere they are created/referenced, but it will work with `ml` of course, but thats why its ml here

woodsp-ibm commented on .github/workflows/neko.yml:
```diff
@@ -0,0 +1,16 @@
+name: Qiskit Neko Integration Tests
```
I have node idea whether anything will be done with this. I can imagine it being removed for apps.

woodsp-ibm commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,340 @@
+
+https://qiskit.org/documentation/contributing_to_qiskit.html
+
+## Contributing to Qiskit Machine Learning
```
```suggestion
## Contributing to Qiskit Algorithms

woodsp-ibm commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,340 @@
+## Contributing to Qiskit Machine Learning
+
+In addition to the general guidelines above there are specific details for
+contributing to Qiskit Machine Learning.
```
```suggestion
contributing to Qiskit Algorithms.
```

woodsp-ibm commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,340 @@
+
+### Project Code Style.
+
+Code in Qiskit Machine Learning should conform to PEP8 and style/lint checks are run to validate
```
```suggestion
Code in Qiskit Algorithms should conform to PEP8 and style/lint checks are run to validate
```

woodsp-ibm commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,340 @@
+please ensure that:
+
+1. The code follows the _code style_ of this project and successfully
+   passes the _unit tests_. Machine Learning uses [Pylint](https://www.pylint.org) and
```
```suggestion
   passes the _unit tests_. Qiskit Algorithms uses [Pylint](https://www.pylint.org) and
```

woodsp-ibm commented on CONTRIBUTING.md:
````diff
@@ -0,0 +1,340 @@
+   make lint
+   make style 
+   ```
+   from the root of the Machine Learning repository clone for lint and style conformance checks.
````
```suggestion
   from the root of the Algorithms repository clone for lint and style conformance checks.
```

woodsp-ibm commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,340 @@
+
+#### Building release notes locally
+
+Building The release notes are part of the standard qiskit-machine-learning documentation
```
```suggestion
Building The release notes are part of the standard qiskit-algorithms documentation
```

woodsp-ibm commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,340 @@
+build all the documentation into `docs/_build/html` and the release notes in
+particular will be located at `docs/_build/html/release_notes.html`
+
+## Installing Qiskit Machine Learning from source
```
```suggestion
## Installing Qiskit Algorithms from source
```

woodsp-ibm commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,340 @@
+
+## Installing Qiskit Machine Learning from source
+
+Please see the [Installing Qiskit Machine Learning from
+Source](https://github.com/qiskit-community/qiskit-machine-learning#installation)
```
```suggestion
Please see the [Installing Qiskit Algorithms from
Source](https://github.com/qiskit-community/qiskit-algorithms#installation)
```

woodsp-ibm commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,340 @@
+Source](https://github.com/qiskit-community/qiskit-machine-learning#installation)
+section of the Qiskit documentation.
+
+Note: Machine Learning depends on Terra, and has optional dependence on Aer and IBM Q Provider, so
```
```suggestion
Note: Qiskit Algorithms depends on Terra, and has optional dependence on Aer, so
```

woodsp-ibm commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,340 @@
+section of the Qiskit documentation.
+
+Note: Machine Learning depends on Terra, and has optional dependence on Aer and IBM Q Provider, so
+these should be installed too. The main branch of Machine Learning is kept working with those other element
```
```suggestion
these should be installed too. The main branch of Algorithms is kept working with those other element
```

woodsp-ibm commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,340 @@
+main branches so these should be installed from source too following the instructions at 
+the same location
+
+Machine Learning also has some other optional dependents see 
+[Machine Learning optional installs](https://github.com/qiskit-community/qiskit-machine-learning#optional-installs) for
```
```suggestion
Algorithms also has some other optional dependents see 
[Algorithms optional installs](https://github.com/qiskit-community/qiskit-algorithms#optional-installs) for
```
I somehow doubt there are any optional installs - I can't think of any at present. I guess the target here can say there are none and this can be kept

woodsp-ibm commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,340 @@
+[**tox**](https://tox.readthedocs.io/en/latest/#). For more information about using tox please
+refer to
+[Terra CONTRIBUTING](https://github.com/Qiskit/qiskit-terra/blob/main/CONTRIBUTING.md#test)
+Test section. However please note Machine Learning does not have any
```
```suggestion
Test section. However please note Algorithms does not have any
```

woodsp-ibm commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,340 @@
+
+### Development Cycle
+
+The development cycle for qiskit-machine-learning is informed by release plans in the 
```
```suggestion
The development cycle for qiskit-algorithms is informed by release plans in the 
```

woodsp-ibm replied:
I have no idea going forwards with this - ie why a community project would have rfcs in Qiskit organization. Same is true for apps.

ElePT replied:
Well, there is at least one, the one about the repo migration

woodsp-ibm replied:
Right but then it was talking about code in the Qiskit org... now, no more. Its not of course all that important in the scheme of things at present - just figured I comment while I saw it.

woodsp-ibm commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,340 @@
+
+* `main`:
+
+The main branch is used for development of the next version of qiskit-machine-learning.
```
```suggestion
The main branch is used for development of the next version of qiskit-algorithms.
```

woodsp-ibm commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,340 @@
+
+### Release Cycle
+
+From time to time, we will release brand new versions of Qiskit Machine Learning. These
```
```suggestion
From time to time, we will release brand new versions of Qiskit Algorithms. These
```

woodsp-ibm commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,340 @@
+
+## Dealing with the git blame ignore list
+
+In the qiskit-machine-learning repository we maintain a list of commits for git blame
```
```suggestion
In the qiskit-algorithms repository we maintain a list of commits for git blame
```

woodsp-ibm commented on CONTRIBUTING.md:
````diff
@@ -0,0 +1,340 @@
+with something like
+
+```
+git blame --ignore-revs-file .git-blame-ignore-revs qiskit_machine_learning/version.py
````
```suggestion
git blame --ignore-revs-file .git-blame-ignore-revs qiskit_algorithms/version.py
```

woodsp-ibm commented on README.md:
```suggestion
Qiskit Algorithms was inspired, authored and brought about by the collective work of a team of researchers.
Algorithms continues to grow with the help and work of
```

woodsp-ibm commented on qiskit_algorithms/version.py:
```diff
@@ -0,0 +1,84 @@
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Contains the Machine Learning version."""
```
```suggestion
"""Contains the Algorithms version."""
```

woodsp-ibm commented on requirements-dev.txt:
```diff
@@ -0,0 +1,19 @@
+coverage>=4.4.0
+matplotlib>=2.1
+seaborn
```
I know ML uses seaborn for tutorials - I doubt it will be needed here

ElePT replied:
```suggestion
```

woodsp-ibm commented on requirements.txt:
```diff
@@ -0,0 +1,7 @@
+scipy>=1.4
+numpy>=1.17
+psutil>=5
+fastdtw
```
Do we use this - time  evolvers?

woodsp-ibm commented on requirements.txt:
```diff
@@ -0,0 +1,7 @@
+psutil>=5
+fastdtw
+setuptools>=40.1.0
+dill>=0.3.4
```
ML needed this - but I think not for algorithms

ElePT replied:
```suggestion
```

woodsp-ibm commented on tools/deploy_documentation.sh:
```diff
@@ -0,0 +1,32 @@
+echo "Pushing built docs to qiskit.org/ecosystem"
+rclone sync --progress --exclude locale/** ./docs/_build/html IBMCOS:qiskit-org-web-resources/ecosystem/algorithms
+
+echo "Pushing built docs to qiskit.org/documentation"
+rclone sync --progress --exclude locale/** ./docs/_build/html IBMCOS:qiskit-org-web-resources/documentation/algorithms
```
We should get rid of this - this is from where the apps pushed to documentation before they were switched to ecosystem. It pushed to both locations - and still does - the original documentation location push ought to be removed from the apps now too.

ElePT replied:
```suggestion
```

ElePT commented on tools/deploy_documentation.sh:
```diff
@@ -0,0 +1,30 @@
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+# Script for pushing the documentation to the qiskit.org/ecosystem and qiskit.org/documentation
```
```suggestion
# Script for pushing the documentation to the qiskit.org/ecosystem
```

ElePT commented on .github/workflows/deploy-docs.yml:
```diff
@@ -0,0 +1,83 @@
+          make html
+          tools/deploy_documentation.sh
+        shell: bash
+  deploy-translatable-strings:
+    if: ${{ startsWith(github.ref, 'refs/heads/stable') && contains('["ElePT","stefan-woerner","woodsp-ibm"]', github.actor) }}
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        python-version: [3.8]
+    steps:
+    - uses: actions/checkout@v3
+      with:
+        fetch-depth: 0
+    - name: Set up Python
+      uses: actions/setup-python@v4
+      with:
+        python-version: ${{ matrix.python-version }}
+    - name: Install dependencies
+      run: |
+        python -m pip install --upgrade pip
+        pip install jupyter qiskit-terra[visualization] tox
+        sudo apt-get install -y graphviz pandoc
+    - name: Build and publish
+      env:
+        encrypted_deploy_po_branch_key: ${{ secrets.encrypted_deploy_po_branch_key }}
+        encrypted_deploy_po_branch_iv: ${{ secrets.encrypted_deploy_po_branch_iv }}
+        QISKIT_PARALLEL: False
+        QISKIT_DOCS_BUILD_TUTORIALS: 'always'
+      run: |
+        echo "earliest_version: 0.1.0" >> releasenotes/config.yaml
+        tools/ignore_untagged_notes.sh
+        tools/deploy_translatable_strings.sh
+      shell: bash
```
```suggestion
```

ElePT commented on .github/workflows/deploy-docs.yml:
```diff
@@ -0,0 +1,83 @@
+
+jobs:
+  docs_publish:
+    if: ${{ startsWith(github.ref, 'refs/heads/stable') && contains('["ElePT","stefan-woerner","woodsp-ibm"]', github.actor) }}
```
```suggestion
    if: ${{ startsWith(github.ref, 'refs/heads/stable') && contains('["ElePT","Cryoris","woodsp-ibm"]', github.actor) }}
```

ElePT replied:
```suggestion
    if: ${{ startsWith(github.ref, 'refs/heads/stable') && contains('["ElePT","Cryoris","woodsp-ibm"]', github.actor) }}
```

ElePT commented on requirements-dev.txt:
```diff
@@ -1,15 +1,8 @@
-sphinx-design>=0.4.0
-sphinxcontrib-spelling
-jupyter-sphinx
 discover
```
```suggestion
discover
rustworkx>=0.13
```

ElePT commented on requirements.txt:
```diff
@@ -1,3 +1,4 @@
 qiskit-terra>=0.24
 scipy>=1.4
 numpy>=1.17
+rustworkx>=0.13
```
```suggestion
```

woodsp-ibm commented on constraints.txt:
```diff
@@ -1 +1,5 @@
+# Numpy 1.25 deprecated some behaviours that we used, and caused the isometry
+# tests to flake. See https://github.com/Qiskit/qiskit-terra/issues/10305,
+# remove pin when resolving that.
+numpy<1.25
```
Is this needed for the unit tests here - while terra may need to constrain things for the unit tests there I imagine it all works ok here or not? If not maybe it needs a bug being raised to fix things.

ElePT replied:
Some unit tests were failing, I don't remember if they were from gradients or optimizers, but this is why I added the constraint.

woodsp-ibm replied:
Ok, we'll have to look into it further then I guess to sort things once this merged.

Eric-Arellano commented on tools/github_poBranch_update_key.enc:
What is this for?

Eric-Arellano commented on tools/generate_spell_dict.py:
Are we sure we  want this? Maybe add it in a follow up?

Eric-Arellano commented on tools/extract_deprecation.py:
What is this for?

woodsp-ibm replied:
The CI has been cloned from ML app. Things can change over eg away from pylintrc going forwards. I will note it seems you never got around to doing the owner run restriction in ML (I noticed it was not here in the copy). Tools like above are used to gather the deprecation messages from the various runs so it was easy to see an overall picture and you did not have to go through each log file.

woodsp-ibm commented on Makefile:
```diff
@@ -0,0 +1,73 @@
+# This code is part of Qiskit.
```
Personally I never use tox - took too much space. Locally with the apps I always run things with make.

Cryoris replied:
Same here, I'm usually using `make` to quickly check things as it runs very quickly. Though that also backfired sometimes if the wrong version of a tool was installed, which you don't get with tox 🤔

ElePT commented:
Thanks @Eric-Arellano for the suggestions! I have added terra's `pyproject.toml` (minus some configuration that seemed not to affect algorithms) and removed `.pylintrc`, and for the tools we are currently not using, I have removed them too. As Steve mentioned, I just copied over a lot of the application modules' files to get a head start setting the repo up (also because it's my first time), but I am with you in trying to keep consistency. It's better to add them as we find them necessary.

Cryoris reviewed (approved):
Re-approving since Eric's suggestion has been added but annulled the approval
